File: cpp/june/day_18_H-Index II.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int N = citations.size(), mid, left = 0, right = N-1;
        while(left <= right){
            mid = left + (right - left)/2;
            if(citations[mid] == N - mid){
                return citations[mid];
            }else if(citations[mid] > N - mid){
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return N - left;
    }
};

File: python3/june/day_18_H-Index II.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        N, left, right = len(citations), 0, len(citations)-1
        while left <= right:
            mid = left + (right - left)//2
            if citations[mid] == N - mid:
                return citations[mid]
            elif citations[mid] > N - mid:
                right = mid - 1
            else:
                left = mid + 1
        return N - left
